### ISSUE TITLE
languagetool is used by malware to exfiltrate credit cards

### ISSUE THREAD

**Issue by mitar** [languagetool-org/languagetool#11091]
TLDR: I think it would be useful to write somewhere that if you find this extension installed on your computer and you didn't install it, you probably have malware installed.

Alternatively, this extension could stop prevent itself from sending credit-card-number-looking data to organization-operated servers.

---

So I found an extension (added to all browsers, Firefox, Chrome, Edge) that I have not installed: A Grammar Checker & Paraphraser LanguageTool. It seems unchanged.

<img width="521" alt="browser-extension" src="https://github.com/user-attachments/assets/0e9d2e1f-04dc-4daa-887d-2061010cbc51">

I wanted to remove it, but I could not. The browser was also configured to be managed by an organization:

<img width="955" alt="browser-is-managed" src="https://github.com/user-attachments/assets/8a004b87-84c5-4cae-9572-53329cf5b676">

And of course it is not. It is a personal device. But malware is making hard to remove it.

So I investigated this further and this extension has a feature that it can be managed and that organization can configure it so that it calls into organization-operated server and not a cloud server managed by the author of the extension. Seems reasonable, organization can prevent leaking company data this way.

On my computer this is configured to localhost. It is also configured to do this for all websites.

<img width="960" alt="extension-sends-data-to-localhost" src="https://github.com/user-attachments/assets/da04da43-2df9-4e5b-b1f2-ef1f25a2c871">

Who listens there? nestat says a Canon printer driver executable? Strange again. That executable seems legitimate (CNAB4RPD.EXE). I suspect it can be exploited so that it runs something else in memory. Opening the port in the browser loads the error that URI handler does not exit for /. It seems to be made with oatpp C++ web framework.

So malware compromises a legitimate printer driver to run a port to which the extension sends all web site form data (including credit card data entered into any web site), which is then send further to their servers. Pretty smart to hide malware this way.

I think this extension could help here a bit by not sending data from password fields and data which looks like a credit card number.

**Comment by tiff** [comment]
Thanks so much for reporting this. We (LanguageTool) are investigating this now. I'll keep you posted.

**Comment by tiff** [comment]
Investigation is ongoing.

However, the title is a bit misleading: LanguageTool only sends texts from multi-line text fields to servers. Password fields and credit card fields are intentionally not handled by the LanguageTool browser extension.

**Comment by mitar** [comment]
That is interesting. So I investigated this after a credit card number was stolen, so I then started investigating how it happened. It might be that also some other fields are sent? Or that a particular website was using a multi-line text field? But I would doubt that malware writers would limit themselves only to such cases? So it might be that more than just multi-line text is send?

**Comment by mitar** [comment]
I doubt malware would rely on people accidentally copying values? In [Google folder](https://drive.google.com/drive/u/0/folders/1l04Y6ozRCv-Eq_3bs0ywivU0W-KfH24F) you can see the copy of the extension I made from the computer with malware. Maybe it is somehow patched/modified?

**Comment by stefanb** [comment]
FYI, the provided Canon printer driver `CNAB4RPD.EXE` has a long history with @virustotal, dating back to 2009/2010:
https://www.virustotal.com/gui/file/58b20cdfe8175005c4174892d9f3811411a7316fe435e3f0ae09ffe52c361831/behavior

**Comment by mitar** [comment]
@stefanb Thanks for the link, but I I am not sure where can I find that long history? I see only that it was analyzed, but I do not see that much has been found? To me it looks like a non-malicious driver which can be compromised (given that it is signed)? Or is it a malicious program just under this name? It does look strange that it is opening ports and connecting to the Internet.

**Comment by Krzysiu** [comment]
> I doubt malware would rely on people accidentally copying values? 

But it does. Many keyloggers also log clipboard, and some are specialized to gather certain data. I don't remember the name, but there was one infamous a few years ago that targeted credit card numbers.
